presto-main/src/main/java/io/prestosql/operator/DynamicFilterCollectOperator.java

```diff
@@ -0,0 +1,207 @@
+
+import static com.google.common.base.Verify.verify;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverByteSize;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxPerDriverValueCount;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This operator acts as a simple "pass-through" pipe, while saving its input pages.
+ * The collected pages' value are used for creating a run-time filtering constraint (for probe-side table scan in an inner join).
+ * We support only small build-side pages (which should be the case when using "broadcast" join).
+ */
+public class DynamicFilterCollectOperator
+        implements Operator
+{
+    public static class DynamicFilterCollectOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final LocalDynamicFilter dynamicFilter;
+        private final Type[] buildTypes;
+
+        public DynamicFilterCollectOperatorFactory(int operatorId, PlanNodeId planNodeId, LocalDynamicFilter dynamicFilter, List<Symbol> buildSymbols, TypeProvider types)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.dynamicFilter = requireNonNull(dynamicFilter, "dynamicFilter is null");
+            requireNonNull(buildSymbols, "buildSymbols is null");
+            requireNonNull(types, "types is null");
+
+            buildTypes = dynamicFilter
+                    .getChannels()
+                    .stream()
+                    .map(channel -> channel.getBuildChannelIndex())
+                    .map(index -> buildSymbols.get(index))
+                    .map(symbol -> types.get(symbol))
+                    .toArray(Type[]::new);
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            return new DynamicFilterCollectOperator(
+                    driverContext.addOperatorContext(operatorId, planNodeId, DynamicFilterCollectOperator.class.getSimpleName()),
+                    dynamicFilter,
+                    buildTypes);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("duplicate() is not supported for DynamicFilterCollectOperatorFactory");
+        }
+    }
+
+    private final OperatorContext context;
+    private boolean finished;
+    private Page current;
+    private final Consumer<TupleDomain<String>> dynamicPredicateConsumer;
+    private final int maxValueCount;
+    private final long maxByteSize;
+
+    private final List<LocalDynamicFilter.Channel> channels;
+    private final Type[] buildTypes;
+    private HashSet<Object>[] valueSets;  // may be dropped if the predicate becomes too large.
```

Comment:
you can use `io.prestosql.operator.aggregation.TypedSet` instead

Reply:
Thanks for the suggestion!
Refactored to use `TypedSet` and a separate `BlockBuilder` for each channel.